Standardized caching options (#629)
Shopify/hydrogen: packages/hydrogen/src/framework/docs/cache.md

| `TenSecondCache()` | `public, max-age=1, stale-while-revalidate=9`             |
| `OneHourCache()`   | `public, max-age=1800, stale-while-revalidate=1800`       |
| `OneDayCache()`    | `public, max-age=3600, stale-while-revalidate=82800`      |
| `OneMonthCache()`  | `public, max-age=1296000, stale-while-revalidate=1296000` |

## Build your own Caching Strategies

If the above caching strategies doesn't work for you, you can create your own caching
strategies that you can use across your project.

For example, you want to create a cache control header of `max-age=30, must-revalidate, no-transform`

```tsx
import type {
  CachingStrategy,
  AllCacheOptions,
} from '@shopify/hydrogen/dist/esnext/types';

export function MustRevalidateNoTransform(
  overrideOptions?: CachingStrategy
): AllCacheOptions {
  return {
    mode: 'must-revalidate, no-transform',
    maxAge: 30,
    ...overrideOptions,
  };
}

// Elsewhere in project
response.cache(MustRevalidateNoTransform());

// Overide option to get `max-age=60, must-revalidate, no-transform`
response.cache(
  MustRevalidateNoTransform({
    maxAge: 60,
  })
);

> ```

Author: that is a good idea

## Caching strategies

Hydrogen includes recommended caching strategies to help you determine which
cache control header to set.

| Caching strategy | Cache control header                                      | Cache duration |
| ---------------- | --------------------------------------------------------- | -------------- |
| `CacheSeconds()` | `public, max-age=1, stale-while-revalidate=9`             | 10 seconds     |
| `CacheMinutes()` | `public, max-age=900, stale-while-revalidate=900`         | 30 minutes     |
| `CacheHours()`   | `public, max-age=1800, stale-while-revalidate=1800`       | 1 hour         |
| `CacheDays()`    | `public, max-age=3600, stale-while-revalidate=82800`      | 1 Day          |
| `CacheWeeks()`   | `public, max-age=604800, stale-while-revalidate=604800`   | 2 Weeks        |
| `CacheMonths()`  | `public, max-age=1296000, stale-while-revalidate=1296000` | 1 Month        |
| `CacheCustom()`  | Define your own cache control header                      | Custom         |

### Example

```jsx
import {CacheSeconds} from '@shopify/hydrogen';
response.cache(CacheSeconds());
```

> ## Build your own caching strategies

Reviewer: ```suggestion
### Build your own caching strategies
```

> Note:
> If you’re interacting with personalized or private data, then you need to override these defaults to meet your needs.

## Caching Strategies

Shopify provides recommended caching strategies to ease the process of determining what
cache control header to set.

Example usage:

```jsx
import {CacheSeconds} from '@shopify/hydrogen';

response.cache(CacheSeconds());
```

> 

Reviewer: ```suggestion
The following table lists the available caching strategies and their associated cache control headers and cache durations:
```

| `CacheDays()`    | `public, max-age=3600, stale-while-revalidate=82800`      | 1 Day          |
| `CacheWeeks()`   | `public, max-age=604800, stale-while-revalidate=604800`   | 2 Weeks        |
| `CacheMonths()`  | `public, max-age=1296000, stale-while-revalidate=1296000` | 1 Month        |
| `CacheCustom()`  | Define your own cache control header                      | Custom         |

## Build your own Caching Strategies

If the above caching strategies doesn't work for you, you can create your own caching
strategies that you can use across your project.

For example, you want to create a cache control header of `max-age=30, must-revalidate, no-transform`

```tsx
response.cache(
  CacheCustom({
    mode: 'must-revalidate, no-transform',
    maxAge: 30,
  })
);
```

### Cache options

Each mechanism accepts the same cache options API based on the [`Cache-Control` HTTP Header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control):

```ts
export interface AllCacheOptions {
  mode?: string;
  maxAge?: number;
  staleWhileRevalidate?: number;
  sMaxAge?: number;
  staleIfError?: number;
}
```

> `mode` option is for setting options that doesn't need a duration
> 
> - `no-store` - The response is prevented from being cached at any layer. This is useful for private or time-sensitive data.
> - `private` - The response is cached in a user’s browser but not at the hosting or edge layer. This is useful for private or customized data.
> - `must-revalidate` - The response must revalidate with server when `max-age` time is expired.

Reviewer: ```suggestion
You can use `mode` to set options that don't need a duration:

- `no-store`: The response is prevented from being cached at any layer. This is useful for private or time-sensitive data.
- `private`: The response is cached in a user’s browser but not at the hosting or edge layer. This is useful for private or customized data.
- `must-revalidate`: The response must revalidate with the server when `max-age` time is expired.
```

> **Note:** There are other available cache control headers but some of them doesn't make sense to be used with Hydrogen. For example,
> `no-cache` option instructs the browser to not use the cached entry until it obtained a 304 (Not Modified) status from server.
> However, Hydrogen server does not send a 304 status on a request.

Reviewer: ```suggestion
> Note: 
> There are other available cache control headers, but some of them aren't applicable to Hydrogen. For example, the `no-cache` option instructs the browser to not use the cached entry until it returns a `304 (Not Modified)` status from server. However, the Hydrogen server doesn't send a 304 status on a request.
```
